File: spec/features/pwb/property_browsing_spec.rb
require 'rails_helper'

module Pwb
  RSpec.describe "Property Browsing", type: :feature do
    # US-1.1: View Property Listings
    # As a public visitor, I want to browse available properties
    # So that I can find properties that interest me

    # Use legacy Prop model which works with existing tests
    let!(:website) { FactoryBot.create(:pwb_website, subdomain: 'test-browse') }
    let!(:sale_prop) do
      ActsAsTenant.with_tenant(website) do
        FactoryBot.create(:pwb_prop, :sale, website: website, title_en: 'Cozy City Apartment')
      end
    end
    let!(:rental_prop) do
      ActsAsTenant.with_tenant(website) do
        FactoryBot.create(:pwb_prop, :long_term_rent, website: website, title_en: 'Downtown Studio')
      end
    end

    before(:each) do
      Capybara.app_host = 'http://test-browse.example.com'
    end

    after(:each) do
      Capybara.app_host = nil
    end

    describe "Sale Properties Page" do
      scenario "displays search filters for sale properties" do
        visit('/en/buy')

        # The buy page should have a search form with price filters
        expect(page).to have_content('Search').or have_content('Filter')
        expect(page).to have_content('Price')
      end

      scenario "can access property details directly" do
        visit("/en/properties/for-sale/#{sale_prop.id}/cozy-city-apartment")

        # Should display property information or a valid page
        expect(page).to have_content('Cozy')
          .or have_content('property')
          .or have_content('Apartment')
          .or have_content('Test Company')
      end
    end

    describe "Rental Properties Page" do
      scenario "displays search filters for rental properties" do
        visit('/en/rent')

        # The rent page should have rental-specific filters
        expect(page).to have_content('Search').or have_content('Filter')
        expect(page).to have_content('Rent')
      end

      scenario "can access rental property details directly" do
        visit("/en/properties/for-rent/#{rental_prop.id}/downtown-studio")

        # Should display rental property information or a valid page
        expect(page).to have_content('Downtown')
          .or have_content('Studio')
          .or have_content('property')
          .or have_content('Test Company')
      end
    end

    describe "Navigation between property types" do
      scenario "can navigate from buy to rent page" do
        visit('/en/buy')

        # Look for rent/rental link in navigation
        if page.has_link?('Rent')
          click_link('Rent')
          expect(current_path).to include('/rent')
        end
      end
    end
  end
end

File: tests/e2e/public/property-browsing.spec.js
// @ts-check
const { test, expect } = require('@playwright/test');
const { TENANTS, ROUTES } = require('../fixtures/test-data');
const { goToTenant, waitForPageLoad } = require('../fixtures/helpers');

/**
 * Property Browsing Tests
 * Migrated from: spec/features/pwb/property_browsing_spec.rb
 *
 * US-1.1: View Property Listings
 * As a public visitor, I want to browse available properties
 * So that I can find properties that interest me
 */

test.describe('Property Browsing', () => {
  const tenant = TENANTS.A;

  test.describe('Sale Properties Page', () => {
    test('displays search filters for sale properties', async ({ page }) => {
      await goToTenant(page, tenant, ROUTES.BUY);

      // The buy page should have a search form with price filters
      const pageContent = await page.content();
      const hasSearchOrFilter = pageContent.includes('Search') || pageContent.includes('Filter');
      const hasPrice = pageContent.includes('Price');

      expect(hasSearchOrFilter).toBeTruthy();
      expect(hasPrice).toBeTruthy();
    });

    test('can access property listing page', async ({ page }) => {
      await goToTenant(page, tenant, ROUTES.BUY);

      // Page should load without errors
      await expect(page).toHaveURL(/\/en\/buy/);

      // Should have some content indicating it's a property listing page
      const pageContent = await page.content();
      const hasPropertyContent = pageContent.includes('Property') ||
                                  pageContent.includes('property') ||
                                  pageContent.includes('Search') ||
                                  pageContent.includes('Filter');
      expect(hasPropertyContent).toBeTruthy();
    });
  });

  test.describe('Rental Properties Page', () => {
    test('displays search filters for rental properties', async ({ page }) => {
      await goToTenant(page, tenant, ROUTES.RENT);

      // The rent page should have rental-specific filters
      const pageContent = await page.content();
      const hasSearchOrFilter = pageContent.includes('Search') || pageContent.includes('Filter');
      const hasRent = pageContent.includes('Rent') || pageContent.includes('rent');

      expect(hasSearchOrFilter).toBeTruthy();
      expect(hasRent).toBeTruthy();
    });

    test('can access rental listing page', async ({ page }) => {
      await goToTenant(page, tenant, ROUTES.RENT);

      // Page should load without errors
      await expect(page).toHaveURL(/\/en\/rent/);

      // Should have some content indicating it's a rental page
      const pageContent = await page.content();
      const hasRentalContent = pageContent.includes('Rent') ||
                                pageContent.includes('rent') ||
                                pageContent.includes('Property');
      expect(hasRentalContent).toBeTruthy();
    });
  });

  test.describe('Navigation between property types', () => {
    test('can navigate from buy to rent page', async ({ page }) => {
      await goToTenant(page, tenant, ROUTES.BUY);

      // Look for rent link specifically in navigation by href
      // Navigation links come from database, so text may vary
      const rentLink = page.locator('nav a[href*="/rent"]').first();
      const count = await rentLink.count();

      if (count > 0) {
        await rentLink.click();
        await waitForPageLoad(page);
        await expect(page).toHaveURL(/\/rent/);
      } else {
        // Skip test if no rent link in navigation (database-dependent)
        test.skip();
      }
    });

    test('can navigate from rent to buy page', async ({ page }) => {
      await goToTenant(page, tenant, ROUTES.RENT);

      // Look for buy link specifically in navigation by href
      // Navigation links come from database, so text may vary
      const buyLink = page.locator('nav a[href*="/buy"]').first();
      const count = await buyLink.count();

      if (count > 0) {
        await buyLink.click();
        await waitForPageLoad(page);
        await expect(page).toHaveURL(/\/buy/);
      } else {
        // Skip test if no buy link in navigation (database-dependent)
        test.skip();
      }
    });
  });
});
